src/lib.rs

```diff
@@ -63,34 +62,39 @@ impl ToOperation for TodoOperation {
     }
 }
 
-pub fn operation_from_string(s: String) -> TodoOperation {
-    let vec: Vec<&str> = s.split('\0').collect();
-
-    TodoOperation {
-        operation: match vec[0] {
-            "add" => OperationType::Add,
-            "remove" => OperationType::Remove,
-            _ => panic!("expected valid operation type, got {}", vec[0]),
-        },
-        todo: TodoData {
-            engine: vec[1].to_string(),
-            rule_id: vec[2].to_string(),
-            file_path: vec[3].to_string(),
-            range: Range {
-                start: Position {
-                    line: i32::from_str_radix(vec[4], 10).expect("valid range.start.line"),
-                    column: i32::from_str_radix(vec[5], 10).expect("valid range.start.column"),
-                },
-                end: Position {
-                    line: i32::from_str_radix(vec[6], 10).expect("valid range.end.line"),
-                    column: i32::from_str_radix(vec[7], 10).expect("valid range.end.column"),
+impl<S> From<S> for TodoOperation
+where
+    S: AsRef<str>,
```

Comment:
Ooohhh boy, Gotta go learn about this one 😸

Reply:
[Docs!](http://doc.rust-lang.org/1.56.0/std/convert/trait.AsRef.html) (Meant to link them before.) It just lets you say "I will take anything you can give me that can do `AsRef<T>`", and the compiler will do the `as_ref` conversion for you. Since it's just converting the reference type, *not* the thing being referenced, it's *super* cheap.

Reply:
Oh, see also [Borrow](http://doc.rust-lang.org/1.56.0/std/borrow/trait.Borrow.html), which is more powerful but also has more restrictions as a result.